File: BackEnd/src/data/IMappers/IWeatherMapper.ts
/**
 * @file IWeatherMapper.ts
 * @description Interfaz del mapper de datos meteorológicos.
 * Convierte entre WeatherCacheDocument (MongoDB), WeatherCache (entidad) y los DTOs de respuesta.
 * @module Weather
 * @layer Data
 */

import type { WeatherCache } from '../../domain/entities/WeatherCache.js';
import type { WeatherCacheDocument } from '../datasources/mongodb/models/WeatherCacheModel.js';
import type { WeatherResponseDTO } from '../../domain/dtos/weather/weather-response.dto.js';
import type { ForecastResponseDTO } from '../../domain/dtos/weather/forecast-response.dto.js';

/**
 * Mapper de WeatherCache: document ↔ entity ↔ DTOs de respuesta.
 */
export interface IWeatherMapper {
  /**
   * Convierte un documento MongoDB en entidad de dominio WeatherCache.
   *
   * @param doc — Documento de la colección weather_cache.
   * @returns WeatherCache
   */
  toEntity(doc: WeatherCacheDocument): WeatherCache;

  /**
   * Convierte una entidad WeatherCache en documento MongoDB.
   * NO incluye _id (se asigna en el repositorio).
   *
   * @param entity — Entidad de dominio.
   * @returns Omit<WeatherCacheDocument, '_id'>
   */
  toDocument(entity: WeatherCache): Omit<WeatherCacheDocument, '_id'>;

  /**
   * Convierte la entidad WeatherCache en DTO de clima actual.
   *
   * @param entity — Entidad de dominio.
   * @returns WeatherResponseDTO
   */
  toWeatherResponseDTO(entity: WeatherCache): WeatherResponseDTO;

  /**
   * Convierte la entidad WeatherCache en DTO de previsión por horas.
   *
   * @param entity — Entidad de dominio.
   * @param hours — Número máximo de horas a incluir.
   * @returns ForecastResponseDTO
   */
  toForecastResponseDTO(entity: WeatherCache, hours?: number): ForecastResponseDTO;
}

File: FrontEnd/lib/data/i_mappers/i_weather_mapper.dart
/// @file i_weather_mapper.dart
/// @description Interfaz del mapper de datos meteorológicos. Contrato WeatherModel ↔ WeatherData.
/// @module Weather
/// @layer Data
library;

import '../../domain/entities/weather_data.dart';
import '../models/weather_model.dart';

/// Contrato de conversión entre el modelo de serialización y la entidad de dominio.
///
/// [injectable] registrar en container.dart como singleton.
abstract interface class IWeatherMapper {
  /// Convierte un modelo de serialización a entidad de dominio.
  WeatherData toEntity(WeatherModel model);

  /// Convierte una entidad de dominio a modelo de serialización.
  WeatherModel toModel(WeatherData entity);
}
